ops/pebble.py

```diff
@@ -1624,7 +1623,7 @@ def _request_raw(
                 message = f'{type(e2).__name__} - {e2}'
             raise APIError(body, code, status, message)
         except urllib.error.URLError as e:
-            if e.errno == errno.ENOENT:
+            if e.args and isinstance(e.args[0], FileNotFoundError):
```

Comment:
Interesting. Did the `e.errno` test not work? In the implementation it is still a subclass of `OSError`: https://github.com/python/cpython/blob/94525a7a4db3a5e82f0e635e21778e8bb31f800a/Lib/urllib/error.py#L19 but yeah, it looks messy.

Reply:
It did not :smile:. This is probably me showing my age, combined with the message that does give the impression that it would be set. In [Python 2](https://github.com/python/cpython/blob/8d21aa21f2cbc6d50aab3f420bb23be1d081dac4/Lib/urllib.py#L219) (maybe early 3?) you got an IOError and could check the errno, e.g.:

```
$ cat /tmp/badsock.py 
from __future__ import print_function
try:
    import urllib.request as urllib
except ImportError:
    import urllib

try:
    urllib.urlopen("file:///foo.sock")
except Exception as e:
    print(type(e), e.errno)
$ pyenv shell 2.7 && python /tmp/badsock.py 
<type 'exceptions.IOError'> 2
$ pyenv shell 3.12 && python /tmp/badsock.py 
<class 'urllib.error.URLError'> None
```

> In the implementation it is still a subclass of OSError

It is in that it inherits from it, but:

> URLError is a sub-type of OSError, but it doesn't share any of the implementation. 

The `__init__` doesn't set `errno` at all, so you get the `None` attribute value from `OSError` at a class level I assume.

Reply:
Aha, interesting. I (sometimes) love Python's super-dynamic nature. :-)